src/vue/directives/tooltip.js

```diff
@@ -0,0 +1,387 @@
+import { globalize } from '@/vue/filters/globalize'
+
+/**
+ * Tooltip directive
+ *
+ * @param {string} [type='hover'] - 'click' or 'hover' to summon tooltip
+ * @param {string} [width='120px'] - width of tooltip, 120 by default
+ * @param {string} position - position of tooltip where it will be placed
+ * @param {string} text - text of tooltip
+ * @param {string} [bgColor='#000'] - background color of tooltip
+ * @param {string} [textColor='#fff'] - text color
+ * @param {string} [fontSize='1.4rem'] - font size of text
+ *
+ */
+export const tooltip = (() => {
+  const COLOR_BLACK = '#000'
+  const COLOR_WHITE = '#fff'
+  const MARGIN_SIZE = 8
+  const PADDING_SIZE = 8
+  const WIDTH = '120px'
+  const HEIGHT = 'auto'
+  const FONT_SIZE = '1.4rem'
+  const DIRECTIONS = {
+    top: 'top',
+    right: 'right',
+    bottom: 'bottom',
+    left: 'left',
+  }
+  let tooltipText
+  let tooltipWrapper
+  let tooltipArrow
+  let targetHeight
+  let twRect
+  let target
+  let targetRect
+  let tooltipArrowRect
+  let position
+  let currentPosition
+  let bgColor
+  let textColor
+  let fontSize
+  let width
+  let height
+  let resizeHandler = null
+  let tooltipOn = false
+  let currentTarget
+  let bindedOpts
+  let opacity = 0
+
+  // update TooltipWrapper client rect
+  function refreshPositions () {
+    twRect = tooltipWrapper.getBoundingClientRect()
+  }
+
+  function alignTooltip (e) {
+    const target = e.target ? e.target : e
+    const targetRect = target.getBoundingClientRect()
+    const appDiv = document.getElementsByClassName('app__main-content')[0]
+    let spaceToRight = window.innerWidth - (targetRect.left + targetRect.width)
+    let rectHeight = tooltipWrapper.offsetHeight + MARGIN_SIZE
+    let spaceToLeft = appDiv.clientWidth - spaceToRight - targetRect.width
+    let spaceToBottom = window.innerHeight - targetRect.bottom
+    // check if there is enough space for tw on left side
+    if (twRect.width > spaceToLeft && position !== DIRECTIONS.right) {
+      let top
+      if (position === DIRECTIONS.top) {
+        top = -(targetRect.height + MARGIN_SIZE + (PADDING_SIZE * 2)) + 'px'
+        tooltipWrapper.style.right = ''
+        tooltipWrapper.style.left = '50%'
+        tooltipWrapper.style.marginLeft = '-' + (twRect.width / 4) + 'px'
+        currentPosition = DIRECTIONS.top
+      } else if (twRect.width < spaceToRight &&
+          position !== DIRECTIONS.bottom) {
+        tooltipWrapper.style.top = '0'
+        tooltipWrapper.style.left = (0 + targetRect.width) + (2 * MARGIN_SIZE) + 'px'
+        currentPosition = DIRECTIONS.right
+      } else {
+        tooltipWrapper.style.top = '0'
+        tooltipWrapper.style.left = ''
+        top = targetRect.height + MARGIN_SIZE + PADDING_SIZE + 'px'
+        currentPosition = DIRECTIONS.bottom
+      }
+      tooltipWrapper.style.transform = `translateY(${top})`
+      refreshPositions()
+    }
+    // check if there is enough space for tw on right side
```

Comment:
Right here I no longer understand what `tw` means. It's not bad to repeat the long name of something in the docs. Badly, when used some abbreviature without any definition of it.

Reply:
Edited